File: modules/python-modern/src/app/changes.py
"""B4: the change stream, on the stdlib HTTP server.

No language lesson is claimed here. Server-sent events on this stack are HTTP plumbing — a
header, a held socket, a flushed line — and `io` belongs to the framework modules, so
claiming it as a Python fundamental would blur the split this module sits on. The stack
serves the endpoint for contract parity and says nothing more.

ThreadingHTTPServer gives each request its own thread, so a held stream is a parked thread.
That is the honest cost of this shape here, and it is the same tradeoff the Java module
makes cheaper with virtual threads.
"""
from __future__ import annotations

import json
import threading

_LOCK = threading.Lock()
_SUBSCRIBERS: list = []


def subscribe(handler) -> None:
    """Hold the connection open until the client drops it."""
    handler.send_response(200)
    handler.send_header("Content-Type", "text/event-stream")
    handler.send_header("Cache-Control", "no-cache")
    handler.send_header("Connection", "keep-alive")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.end_headers()

    stream = handler.wfile
    with _LOCK:
        _SUBSCRIBERS.append(stream)
    try:
        stream.write(b": open\n\n")
        stream.flush()
        stop = threading.Event()
        while not stop.wait(15):
            stream.write(b": keep-alive\n\n")
            stream.flush()
    except (BrokenPipeError, ConnectionResetError, ValueError):
        pass
    finally:
        with _LOCK:
            if stream in _SUBSCRIBERS:
                _SUBSCRIBERS.remove(stream)


def emit(kind: str, item_id: str) -> None:
    """Announce a change. A dead subscriber is dropped rather than retried."""
    frame = f"event: change\ndata: {json.dumps({'kind': kind, 'id': item_id})}\n\n".encode()
    with _LOCK:
        for stream in list(_SUBSCRIBERS):
            try:
                stream.write(frame)
                stream.flush()
            except (BrokenPipeError, ConnectionResetError, ValueError):
                _SUBSCRIBERS.remove(stream)

File: modules/java-modern/src/main/java/com/abyss/polyglot/javamodern/Changes.java
package com.abyss.polyglot.javamodern;

import com.sun.net.httpserver.HttpExchange;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * B4: the change stream, on the JDK HTTP server and nothing else.
 *
 * No language lesson is claimed here. Server-sent events on this stack are HTTP plumbing
 * (write a header, keep the socket open, flush a line), and `io` belongs to the framework
 * modules, so claiming a language fundamental for it would blur the split this module's
 * side of the demo depends on. The stack serves the endpoint for contract parity and says
 * nothing more.
 *
 * Each subscriber is a held response body. A virtual thread per request is what makes that
 * affordable: the server's executor is newVirtualThreadPerTaskExecutor, so a parked stream
 * costs a heap object rather than an OS thread, and holding a hundred of them is
 * unremarkable.
 */
final class Changes {

    private static final List<OutputStream> SUBSCRIBERS = new CopyOnWriteArrayList<>();

    private Changes() {
    }

    /** Hold the connection open and register it. The handler never returns until it drops. */
    static void subscribe(HttpExchange exchange) throws IOException {
        exchange.getResponseHeaders().add("Content-Type", "text/event-stream");
        exchange.getResponseHeaders().add("Cache-Control", "no-cache");
        exchange.getResponseHeaders().add("Connection", "keep-alive");
        exchange.getResponseHeaders().add("Access-Control-Allow-Origin", "*");
        exchange.sendResponseHeaders(200, 0);

        OutputStream body = exchange.getResponseBody();
        SUBSCRIBERS.add(body);
        try {
            // A comment line on connect, so a client knows the stream is live before the
            // first real change rather than after it.
            body.write(": open\n\n".getBytes(StandardCharsets.UTF_8));
            body.flush();
            while (!Thread.currentThread().isInterrupted()) {
                Thread.sleep(15_000);
                body.write(": keep-alive\n\n".getBytes(StandardCharsets.UTF_8));
                body.flush();
            }
        } catch (IOException | InterruptedException closed) {
            Thread.currentThread().interrupt();
        } finally {
            SUBSCRIBERS.remove(body);
            exchange.close();
        }
    }

    /** Announce a change. A dead subscriber is dropped rather than retried. */
    static void emit(String kind, String id) {
        String frame = "event: change\ndata: {\"kind\":\"%s\",\"id\":\"%s\"}\n\n"
                .formatted(kind, id);
        byte[] bytes = frame.getBytes(StandardCharsets.UTF_8);
        for (OutputStream subscriber : SUBSCRIBERS) {
            try {
                subscriber.write(bytes);
                subscriber.flush();
            } catch (IOException gone) {
                SUBSCRIBERS.remove(subscriber);
            }
        }
    }
}
